stackhpc_ipa_hardware_managers/system_nic.py

```diff
@@ -0,0 +1,241 @@
+            "lspci -vmmnD {filter}".format(filter=device_filter_arg), **kwargs)
+    except (processutils.ProcessExecutionError, OSError):
+        raise errors.CleaningError("Unable to run lspci")
+    return lspci_output
+
+
+def _parse_lspci_output(lspci_output):
+    all_records = lspci_output.split("\n\n")
+    devices = []
+    for record_str in all_records:
+        if record_str == "":
+            continue
+        fields_raw = record_str.splitlines()
+        fields = map((lambda x: x.split(':\t')), fields_raw)
+        devices.append(dict(list(map(tuple, fields))))
+    return devices
+
+
+def _get_device_filter_string(vendor_id, device_id):
+    device_id_str = ""
+    device_id_str += vendor_id or ""
+    device_id_str += ":"
+    device_id_str += device_id or ""
+    return device_id_str
+
+
+def _get_base_in_relative_path(path):
+    # given relative path, returns the base element, e.g one/two/three,
+    # returns one
+    if os.path.isabs(path):
+        raise ValueError("path must be relative")
+    dirname = os.path.dirname(path)
+    if dirname == "":
+        return path
+    else:
+        return _get_base_in_relative_path(dirname)
+
+
+def _pci_addr_to_net_interface(pci_addr):
+    for uevent_file in glob.glob("/sys/class/net/*/device/uevent"):
+        with open(uevent_file, 'r') as f:
+            for line in f.readlines():
+                if _is_matching_uevent_line(line, pci_addr):
+                    return _get_NIC_name(uevent_file)
+
+
+def _is_matching_uevent_line(line, pci_addr):
+    if line.startswith(UEVENT_PCI_SLOT_NAME_PREFIX):
+        # example : "PCI_SLOT_NAME=0000:00:19.0\n"
+        slot_name = line.lstrip(UEVENT_PCI_SLOT_NAME_PREFIX).strip()
+        if slot_name == pci_addr:
+            return True
+    return False
+
+
+def _get_NIC_name(uevent_file):
+    relative_path = os.path.relpath(uevent_file, '/sys/class/net/')
+    return _get_base_in_relative_path(relative_path)
+
+
+def _get_ethtool_output(dev, **kwargs):
+    kwargs["shell"] = True
+    ethtool_output = ''
+    try:
+        ethtool_output, _e = utils.execute(
+            "ethtool -i {device}".format(device=dev),
+            **kwargs)
+    except (processutils.ProcessExecutionError, OSError):
+        raise errors.CleaningError("Unable to run ethtool")
+    return ethtool_output
+
+
+def _get_ethtool_field(ethtool_output, field):
+    for line in ethtool_output.splitlines():
+        (candidate_field, value) = tuple(line.split(": "))
+        if candidate_field == field:
+            return value
+
+
+def _get_expected_property(node, node_property):
+    try:
+        expected_property = node[node_property]
+    except KeyError:
+        raise errors.CleaningError(
+            "Expected property '{0}' not found. You should make sure all items"
+            " in the nic_firmware list contain {0}".format(node_property))
+    return expected_property
+
+
+def _get_nic_firmware_versions(vendor_id, device_id):
```

Comment:
Could use a note describing the returned data format. Looks like it's a dict mapping interface name to firmware version.